### PR TITLE
Adding support for tree filters

### PR DESCRIPTION
**Pull request by HeyZoos**
In this PR I've added support for a JSON Patch-like syntax which allows users to make more complicated queries, specifically with dataset column conditions. There are a lot of changes, and while I'm enthusiastic about that I realize it makes this really tough to review. Instead of listing the changes here, I'm going to comment up all the files -- hopefully this makes things easier. In the future I'll keep the the PRs smaller in scale.

### CODE REVIEW

File: plenario/api/condition_builder.py [hunk 1]
```diff
@@ -20,80 +21,78 @@
 }
 
 
-def parse_general(table, field, value):
```
Comment (HeyZoos) [inline plenario/api/condition_builder.py]: In order to streamline the process of specifying column conditions, I've chosen to convert a user's request args into a dataset__filter (if they didn't choose to include one already). This change removes the need for two different ways to parse user arguments, and allowed me to remove parse_general.

File: plenario/api/condition_builder.py [hunk 2]
```diff
@@ -20,80 +21,78 @@
...
-    """To parse the non-tree style way of providing arguments.
+def parse_tree(table, condition_tree, literally=False):
```
Comment (HeyZoos) [inline plenario/api/condition_builder.py]: Replacing parse_general is parse_tree, which takes a dictionary of the form {"op": "<sqlalchemy_op>", "col": "<dataset_col>", "val": "<target_value>"} and builds a single, cumulative SQLAlchemy condition from it. Because of the recursive nature of the method, it allows one to nest conditions to express more complicated queries such as (FOO1 = BAR1 AND FOO2 = BAR2) OR (FOO3 >= BAR3)

File: plenario/api/condition_builder.py [hunk 3]
```diff
@@ -20,80 +21,78 @@
...
-        return getattr(column, field_ops[operator])(operand)
-
-
-# general_filters
-# ===============
-# Filters which apply to all the tables, but whose columns are not directly
-# specified by an argument key. ex: 'obs_date__ge' translates into a condition
-# that needs to be made for a table's 'point_date' column.
-
-general_filters = {
```
Comment (HeyZoos) [inline plenario/api/condition_builder.py]: Since these conversions were handled exclusively by `parse_general`, they are also not needed.

File: tests/points/validator_tests.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+    def test_validator_with_good_column_but_bad_value(self):
+        endpoint = 'detail'
+        query = '?dataset_name=crimes&data_type=json&id="break_me"'
+
+        resp_data = self.get_json_response_data(endpoint + query)
+
+        self.assertTrue(len(resp_data['meta']['message']) == 2)
+        self.assertTrue('Unused param' in resp_data['meta']['message'][0])
+        self.assertTrue('not a valid value' in resp_data['meta']['message'][1])
```
Comment (WillEngler) [inline tests/points/validator_tests.py]: I don't think we should check for the exact text of error messages in our tests. Making a more descriptive error message would break the test. I prefer to check for HTTP status code.

File: plenario/api/condition_builder.py [hunk 5]
```diff
@@ -20,80 +21,78 @@
...
-def date__time_of_day_filter(table, op, val):
-    """Because I couldn't fit it into a one line lambda.
-
-    :param table: SQLAlchemy table object
-    :param op: string op code
-    :param val: target value to be compared against
+        condition = getattr(column, field_ops[operator])(operand)
 
-    :returns: table condition for just the hour"""
+    if literally:
```
Comment (HeyZoos) [inline plenario/api/condition_builder.py]: This exists specifically for the `/shapes` endpoint, specifically the `export_shape` method. This method relies on a subprocess call made to ogr2ogr, which uses a literal string SQL query to fetch data from postgres. These lines here enable `_operator_to_condition` to return the condition together as a string with its target values, instead of a SQLAlchemy condition with placeholders

File: plenario/api/point.py [hunk 6]
```diff
@@ -9,60 +9,18 @@
...
 from plenario.api.response import form_geojson_detail_response, add_geojson_feature
-from plenario.api.validator import Validator, DatasetRequiredValidator, NoGeoJSONDatasetRequiredValidator
-from plenario.api.validator import NoDefaultDatesValidator, validate, NoGeoJSONValidator
+from plenario.api.validator import DatasetRequiredValidator, NoGeoJSONDatasetRequiredValidator
+from plenario.api.validator import NoDefaultDatesValidator, validate, NoGeoJSONValidator, has_tree_filters
 from plenario.database import session
 from plenario.models import MetaTable
 
 
-def form_detail_sql_query(args, aggregate_points=False):
```
Comment (HeyZoos) [inline plenario/api/point.py]: replaced with `detail_query`

File: plenario/models.py [hunk 7]
```diff
@@ -331,10 +333,9 @@ def timeseries(self, agg_unit, start, end, geom=None, column_filters=None):
         # inspired this implementation.
         t = self.point_table
 
-        if agg_unit == 'quarter':
-            step = '3 months'
-        else:
-            step = '1 ' + agg_unit
+        # Special case for the 'quarter' unit of aggregation.
+        step = '3 months' if agg_unit == 'quarter' else '1 ' + agg_unit
```
Comment (WillEngler) [inline plenario/models.py]: I dig it when you make things prettier like that.

File: plenario/api/point.py [hunk 8]
```diff
@@ -88,19 +46,20 @@ def detail_aggregate():
...
     return _detail_aggregate(validated_args)
 
 
 @cache.cached(timeout=CACHE_TIMEOUT, key_prefix=make_cache_key)
 @crossdomain(origin="*")
 def detail():
     fields = ('location_geom__within', 'dataset_name', 'shape', 'obs_date__ge',
-              'obs_date__le', 'data_type', 'offset')
-    validator = NoGeoJSONDatasetRequiredValidator(only=fields)
+              'obs_date__le', 'data_type', 'offset', 'date__time_of_day_ge',
```
Comment (HeyZoos) [inline plenario/api/point.py]: /detail was not considering hour params because I had left date__time_of_day out of this tuple

File: plenario/api/point.py [hunk 9]
```diff
@@ -110,9 +69,9 @@ def detail():
 def grid():
     fields = ('dataset_name', 'resolution', 'buffer', 'obs_date__le', 'obs_date__ge',
               'location_geom__within')
-    validated_args = validate(DatasetRequiredValidator(), request.args.to_dict())
+    validated_args = validate(DatasetRequiredValidator(only=fields), request.args.to_dict())
```
Comment (HeyZoos) [inline plenario/api/point.py]: /grid was not restricted to just checking fields, and would check any field listed in the DatasetRequiredValidator

File: plenario/models.py [hunk 10]
```diff
@@ -464,6 +467,10 @@ def index(cls, geom=None):
 
         return listing
 
+    @classmethod
```
Comment (WillEngler) [inline plenario/models.py]: Why is this a class method? Looks like we don't use `cls`. Unless you wanted to sub in `cls` for `ShapeMetadata` in `ShapeMetadata.dataset_name`

File: plenario/api/point.py [hunk 11]
```diff
@@ -154,17 +113,28 @@ def meta():
...
-    if dataset is not None:
-        table_names = [dataset.name]
-        # for the query's meta information, so that it doesn't show the index
+    meta_params = ['geom', 'dataset', 'dataset_name__in', 'obs_date__ge', 'obs_date__le', 'agg']
+    meta_vals = [args.data.get(k) for k in meta_params]
+    geom, dataset, table_names, start_date, end_date, agg = meta_vals
+
+    ctrees = {}
+
+    if has_tree_filters(args.data):
```
Comment (HeyZoos) [inline plenario/api/point.py]: builds dataset specific conditions to be applied in the `timseries_all` call if `__filter` params are provided

File: plenario/api/condition_builder.py [hunk 12]
```diff
@@ -20,80 +21,78 @@
...
+        raise ValueError('{} caused parse to fail for table {} with args {}'
+                         .format(ex, table, condition_tree))
 
-    field_tokens = field.split("__")
-    column = table.columns.get(field_tokens[0])
 
-    # equality filters based on table columns
-    if len(field_tokens) == 1 and column is not None:
-        return _operator_to_condition(column, 'eq', value)
+def _parse_condition_tree(table, ctree, literally=False):
```
Comment (WillEngler) [inline plenario/api/condition_builder.py]: I think this came out so nicely. It does more, but it's more readable than our old style of parsing.

File: plenario/api/point.py [hunk 13]
```diff
@@ -154,17 +113,28 @@ def meta():
...
+            if 'filter' in field:
+                metarecord = MetaTable.get_by_dataset_name(field.split('__')[0])
+                pt = metarecord.point_table
+                ctrees[pt.name] = parse_tree(pt, value)
+        # Just cleanliness, since we don't use this argument. Doesn't have
+        # to show up in the JSON response.
+        del args.data['dataset']
+
+    # If a single dataset was provided, it's the only thing we need to consider.
+    if ctrees or dataset is not None:
```
Comment (HeyZoos) [inline plenario/api/point.py]: incorrect condition, when I'm through commenting stuff up I'll fix this

File: plenario/api/condition_builder.py [hunk 14]
```diff
@@ -20,80 +21,78 @@
...
+            _parse_condition_tree(table, child, literally)
+            for child in ctree['val']
+        )
+
+    elif op in field_ops:
+        col = ctree['col']
+        val = ctree['val']
+        try:
+            return _operator_to_condition(
+                getattr(table, 'columns')[col], op, val, literally
```
Comment (WillEngler) [inline plenario/api/condition_builder.py]: Doing a `getattr` when we know the name of the attr feels weird

File: plenario/api/condition_builder.py [hunk 15]
```diff
@@ -20,80 +21,78 @@
...
+            )
+
+        # Exists for date__time_of_day. Since it doesn't come
```
Comment (WillEngler) [inline plenario/api/condition_builder.py]: > Since it doesn't come

Did you trail off there?

File: plenario/api/point.py (hunk 8 above)
Comment (WillEngler) [inline plenario/api/point.py]: We should give that feature some love in the API docs revamp

File: plenario/api/point.py (hunk 11 above)
Comment (WillEngler) [inline plenario/api/point.py]: That is so cool. Did not have this application of the new filter type in mind when I proposed the project.

File: plenario/api/point.py [hunk 16]
```diff
@@ -280,66 +263,128 @@ def _detail_aggregate(args):
...
 
-    elif datatype == 'csv':
-        return form_csv_detail_response(to_remove, rows)
+    elif data_type == 'geojson':
+        return form_geojson_detail_response(to_remove, args, result_rows)
 
-    elif datatype == 'geojson':
-        return form_geojson_detail_response(to_remove, args, rows)
+
+def detail_query(args, aggregate=False):
```
Comment (HeyZoos) [inline plenario/api/point.py]: used by `_detail` and by `_aggregate_point_data`

File: plenario/api/shape.py [hunk 17]
```diff
@@ -1,76 +1,48 @@
...
 
+def _shape_format_to_file_extension(requested_format):
+    format_map = {
+        'json': 'json',
+        'kml': 'kml',
+        'shapefile': 'zip'
+    }
+    return format_map[requested_format]
 
-def make_sql_ready_geom(geojson_str):
```
Comment (HeyZoos) [inline plenario/api/shape.py]: this method was only ever called once in `_get_all_shape_datasets`, I figured I'd just delete it

File: plenario/api/condition_builder.py (hunk 5 above)
Comment (WillEngler) [inline plenario/api/condition_builder.py]: There is a lot of knowledge of how psycopg2 and SQLAlchemy work packed into these lines. Can you add some comments on what's going on? Today me understands this, but 6 months from now me will need some help.

File: plenario/api/shape.py [hunk 18]
```diff
@@ -109,93 +83,143 @@ def get_all_shape_datasets():
...
-        params['shape'] = polygon_dataset_name
-    params['dataset_name'] = point_dataset_name
-
-    args = validate(DatasetRequiredValidator(), params)
-    if args.errors:
-        return bad_request(args.errors)
-
-    # Apply standard filters to point dataset
-    # And join each point to the containing shape
-    q = form_detail_sql_query(args, True)
```
Comment (HeyZoos) [inline plenario/api/shape.py]: this work is moved into the `_aggregate_point_data` route logic

File: plenario/api/shape.py [hunk 19]
```diff
@@ -109,93 +83,143 @@ def get_all_shape_datasets():
...
-    if geom:
-        intersection = dataset.c.geom.ST_Within(
-            func.ST_GeomFromGeoJSON(geom)
-        )
-        q = q.filter(intersection)
-
-    # Page in RESPONSE_LIMIT chunks
-    # This seems contradictory. Don't we want one row per shape, no matter what?
-    offset = args.data['offset']
-    q = q.limit(RESPONSE_LIMIT)
```
Comment (HeyZoos) [inline plenario/api/shape.py]: shapes does not support the ability to limit aggregate queries anymore

File: plenario/api/shape.py [hunk 20]
```diff
@@ -109,93 +83,143 @@ def get_all_shape_datasets():
...
-def _shape_format_to_file_extension(requested_format):
-    format_map = {
-        'json': 'json',
-        'kml': 'kml',
-        'shapefile': 'zip'
-    }
-    return format_map[requested_format]
\ No newline at end of file
+
+def _export_dataset_to_response(shapeset, data_type, query=None):
```
Comment (HeyZoos) [inline plenario/api/shape.py]: No actual changes, just moved it.

File: plenario/api/validator.py [hunk 21]
```diff
@@ -1,15 +1,27 @@
...
+from marshmallow.validate import Range, Length, OneOf, ValidationError
+from sqlalchemy.exc import DatabaseError, ProgrammingError, NoSuchTableError
 
 from plenario.api.common import extract_first_geometry_fragment, make_fragment_str
+from plenario.api.condition_builder import field_ops
 from plenario.database import session
 from plenario.models import MetaTable, ShapeMetadata
 
 
+def validate_many_datasets(list_of_datasets):
```
Comment (HeyZoos) [inline plenario/api/validator.py]: dataset_name__in now get's checked for every table it holds. Originally it was just checked to be a list

File: plenario/api/validator.py [hunk 22]
```diff
@@ -54,13 +66,16 @@ class NoGeoJSONValidator(Validator):
     response format."""
 
     valid_formats = {'csv', 'json'}
+    # Validator re-initialized so that it doesn't use old valid_formats.
+    data_type = fields.Str(default='json', validate=OneOf(valid_formats))
```
Comment (HeyZoos) [inline plenario/api/validator.py]: validator wasn't checking the formats correctly, because the data_type validator would be initialized with the old set of valid_formats. You could still give it 'geojson'

File: plenario/api/validator.py [hunk 23]
```diff
@@ -71,6 +86,13 @@ class NoDefaultDatesValidator(Validator):
     obs_date__le = fields.Date(default=None)
 
 
+class ExportFormatsValidator(Validator):
```
Comment (HeyZoos) [inline plenario/api/validator.py]: special validator for `export_shape`

File: plenario/api/validator.py [hunk 24]
```diff
@@ -131,36 +158,178 @@ def validate(validator, request_args):
...
 
-    # convert string values to corresponding types so the validator can work
-    convert(args)
-    # returns validated parameters as strings
-    result = validator.dump(args)
-    # convert any remaining strings in args to the right type
-    convert(result.data)
+    # For validator dataset_name__in... need to find a better way to
+    # make it play nice with the validator.
+    if args.get('dataset_name__in'):
```
Comment (HeyZoos) [inline plenario/api/validator.py]: dataset_name__in needs to hold a list, otherwise the validator complains

File: plenario/api/validator.py [hunk 25]
```diff
@@ -131,36 +158,178 @@ def validate(validator, request_args):
...
-    unchecked = set(args.keys()) - set(validator.fields.keys())
+    # This first validation step covers conditions that are dataset
+    # agnostic. These are values can be used to apply to all datasets
+    # (ex. obs_date), or concern the format of the response (ex. limit,
+    # datatype, offset).
+
+    # If there are errors, fail quickly and return.
+    result = validator.load(args)
+    if result.errors:
+        return result
```
Comment (HeyZoos) [inline plenario/api/validator.py]: the first `.convert` I used was redundant. The `.load` method converts the arguments, and has the added benefit of reporting errors early, that the we don't need to continue in the validate function if we don't have to

File: plenario/api/validator.py [hunk 26]
```diff
@@ -131,36 +158,178 @@ def validate(validator, request_args):
...
+                        table = ShapeMetadata.get_by_dataset_name(t_name).shape_table
+                    except (AttributeError, NoSuchTableError):
+                        result.errors[t_name] = "Table name {} could not be found.".format(t_name)
+                        return result
+
+                # Report a tree which causes the JSON parser to fail.
+                # Or a tree whose value is not valid.
+                try:
+                    cond_tree = json.loads(value)
+                    if valid_tree(table, cond_tree):
```
Comment (HeyZoos) [inline plenario/api/validator.py]: where the tree is validated

File: plenario/api/validator.py [hunk 27]
```diff
@@ -131,36 +158,178 @@ def validate(validator, request_args):
...
+                t_name = key.split('__')[0]
+
+                # Report a filter which specifies a non-existent tree.
+                try:
+                    table = MetaTable.get_by_dataset_name(t_name).point_table
+                except (AttributeError, NoSuchTableError):
+                    try:
+                        table = ShapeMetadata.get_by_dataset_name(t_name).shape_table
+                    except (AttributeError, NoSuchTableError):
+                        result.errors[t_name] = "Table name {} could not be found.".format(t_name)
```
Comment (HeyZoos) [inline plenario/api/validator.py]: the check here ensures the table exists, before validating the tree

File: plenario/models.py [hunk 28]
```diff
@@ -1,23 +1,21 @@
...
+from sqlalchemy import Table, select, Integer
+from sqlalchemy.dialects.postgresql import ARRAY, JSONB
+from sqlalchemy.orm import synonym
+from uuid import uuid4
 
 from plenario.database import session, Base
-from plenario.utils.helpers import slugify
-
-from collections import namedtuple
+from plenario.utils.helpers import get_size_in_degrees, slugify
```
Comment (HeyZoos) [inline plenario/models.py]: just rearranging the imports -- "a foolish consistency..." I know, but sometimes I just have urges to alphabetize

File: plenario/models.py [hunk 29]
```diff
@@ -204,12 +202,16 @@ def attach_metadata(cls, rows):
     # {'dataset_name': 'Foo',
     # 'items': [{'datetime': dt, 'count': int}, ...] } ]
     @classmethod
-    def timeseries_all(cls, table_names, agg_unit, start, end, geom=None):
+    def timeseries_all(cls, table_names, agg_unit, start, end, geom=None, ctrees=None):
```
Comment (HeyZoos) [inline plenario/models.py]: allows timeseries to support dataset specific column conditions!

File: plenario/models.py [hunk 30]
```diff
@@ -464,6 +467,10 @@ def index(cls, geom=None):
 
         return listing
 
+    @classmethod
+    def tablenames(cls):
```
Comment (HeyZoos) [inline plenario/models.py]: the equivalent to MetaTable's `index` method. Exists because ShapeTable's `index` does something else

**Comment by WillEngler** [comment]
Stray thoughts:

* The cornerstone of this, `_parse_condition_tree ` might be the most elegant piece of code in Plenario. It's really neat.
* I'm happy to see all of the test cases you added, especially the unhappy paths in the validator tests. Rolling out something algorithmic like this is nervewracking because it's hard to come up with all the edge cases. If we start seeing problems in production, we might consider pulling out [the big guns](https://hypothesis.readthedocs.io/en/latest/) for testing this kind of thing.

**Comment by WillEngler** [comment]
Many thanks to @ayang7  for getting this feature started.
